In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt


In [2]:
pd.set_option("display.max_rows",1500)
pd.set_option("display.max_columns",90)

In [3]:
df = pd.read_csv(r"C:\HousePrice_prediction\train.csv")
df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1         Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3         Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4         Lvl    AllPub       FR2       Gtl      NoRidge       Norm   

  Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0       Norm     1Fam     2Story            7            5       2003   
1       Norm     1Fam     1Story            6            8       1976   
2       Norm     1Fam     2Story            7            5       2001   
3       Norm     1Fam     2Story            7            5       1915   
4       Norm     1Fam     2Story            8            5       2000   

   YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0          2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1          1976     Gable  CompShg     MetalSd     MetalSd        NaN   
2          2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
3          1970     Gable  CompShg     Wd Sdng     Wd Shng        NaN   
4          2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   

   MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
0       196.0        Gd        TA      PConc       Gd       TA           No   
1         0.0        TA        TA     CBlock       Gd       TA           Gd   
2       162.0        Gd        TA      PConc       Gd       TA           Mn   
3         0.0        TA        TA     BrkTil       TA       Gd           No   
4       350.0        Gd        TA      PConc       Gd       TA           Av   

  BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  \
0          GLQ         706          Unf           0        150          856   
1          ALQ         978          Unf           0        284         1262   
2          GLQ         486          Unf           0        434          920   
3          ALQ         216          Unf           0        540          756   
4          GLQ         655          Unf           0        490         1145   

  Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  \
0    GasA        Ex          Y      SBrkr       856       854             0   
1    GasA        Ex          Y      SBrkr      1262         0             0   
2    GasA        Ex          Y      SBrkr       920       866             0   
3    GasA        Gd          Y      SBrkr       961       756             0   
4    GasA        Ex          Y      SBrkr      1145      1053             0   

   GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  \
0       1710             1             0         2         1             3   
1       1262             0             1         2         0             3   
2       1786             1             0         2         1             3   
3       1717             1             0         1         0             3   
4       2198             1             0         2         1             4   

   KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu  \
0             1          Gd             8        Typ           0         NaN   
1             1          TA             6        Typ           1          TA   
2             1          Gd             6        Typ           1          TA   
3             1          Gd            

In [4]:
df.shape

(1460, 81)

In [5]:
df.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726   
std       1.112799    30.202904     20.645407   181.066207   456.098091   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000   

        BsmtFinSF2    BsmtUnfSF  TotalBsmtSF     1stFlrSF     2ndFlrSF  \
count  1460.000000  1460.000000  1460.000000  1460.000000  1460.000000   
mean     46.549315   567.240411  1057.429452  1162.626712   346.992466   
std     161.319273   441.866955   438.705324   386.587738   436.528436   
min       0.000000     0.000000     0.000000   334.000000     0.000000   
25%       0.000000   223.000000   795.750000   882.000000     0.000000   
50%       0.000000   477.500000   991.500000  1087.000000     0.000000   
75%       0.000000   808.000000  1298.250000  1391.250000   728.000000   
max    1474.000000  2336.000000  6110.000000  4692.000000  2065.000000   

       LowQualFinSF    GrLivArea  BsmtFullBath  BsmtHalfBath     FullBath  \
count   1460.000000  1460.000000   1460.000000   1460.000000  1460.000000   
mean       5.844521  1515.463699      0.425342      0.057534     1.565068   
std       48.623081   525.480383      0.518911      0.238753     0.550916   
min        0.000000   334.000000      0.000000      0.000000     0.000000   
25%        0.000000  1129.500000      0.000000      0.000000     1.000000   
50%        0.000000  1464.000000      0.000000      0.000000     2.000000   
75%        0.000000  1776.750000      1.000000      0.000000     2.000000   
max      572.000000  5642.000000      3.000000      2.000000     3.000000   

          HalfBath  BedroomAbvGr  KitchenAbvGr  TotRmsAbvGrd   Fireplaces  \
count  1460.000000   1460.000000   1460.000000   1460.000000  1460.000000   
mean      0.382877      2.866438      1.046575      6.517808     0.613014   
std       0.502885      0.815778      0.220338      1.625393     0.644666   
min       0.000000      0.000000      0.000000      2.000000     0.000000   
25%       0.000000      2.000000      1.000000      5.000000     0.000000   
50%       0.000000      3.000000      1.000000      6.000000     1.000000   
75%       1.000000      3.000000      1.000000      7.000000     1.000000   
max       2.000000      8.000000      3.000000     14.000000     3.000000   

       GarageYrBlt   GarageCars   GarageArea   WoodDeckSF  OpenPorchSF  \
count  1379.000000  1460.000000  1460.000000  1460.000000  1460.000000   
mean   1978.506164     1.767123   472.980137    94.244521    46.660274   
std      24.689725     0.747315   213.804841   125.338794    66.256028   
min    1900.000000     0.000000     0.000000     0.000000     0.000000   
25%    1961.000000     1.000000   334.500000     0.000000     0.000000   
50%    1980.000000     2.000000   480.000000     0.000000    25.000000   
75%    2002.000000     2.000000   576.000000   168.000000    68.000000   
max    2

#observation 1 
outliers present in 
mssubclass
lotfrontage
lotarea
masvnrarea
bsmtfinsf1
BsmtFinSF2
BsmtUnfSF
TotalBsmtSF
1stFlrSF	
2ndFlrSF
LowQualFinSF
GrLivArea
BsmtHalfBath
TotRmsAbvGrd
GarageArea
WoodDeckSF
OpenPorchSF	
EnclosedPorch	
3SsnPorch	
ScreenPorch	
PoolArea
MiscVal
SalePrice

In [6]:
df.isnull().sum()

Id                  0
MSSubClass          0
MSZoning            0
LotFrontage       259
LotArea             0
Street              0
Alley            1369
LotShape            0
LandContour         0
Utilities           0
LotConfig           0
LandSlope           0
Neighborhood        0
Condition1          0
Condition2          0
BldgType            0
HouseStyle          0
OverallQual         0
OverallCond         0
YearBuilt           0
YearRemodAdd        0
RoofStyle           0
RoofMatl            0
Exterior1st         0
Exterior2nd         0
MasVnrType        872
MasVnrArea          8
ExterQual           0
ExterCond           0
Foundation          0
BsmtQual           37
BsmtCond           37
BsmtExposure       38
BsmtFinType1       37
BsmtFinSF1          0
BsmtFinType2       38
BsmtFinSF2          0
BsmtUnfSF           0
TotalBsmtSF         0
Heating             0
HeatingQC           0
CentralAir          0
Electrical          1
1stFlrSF            0
2ndFlrSF            0
LowQualFin

#observation 2 
dropping theses features will be better as more than 50 % data is missing
Alley            1369
MasVnrType        872
PoolQC           1453
Fence            1179
MiscFeature      1406

In [7]:
#handeling LotFrontage       259null values by filling mean value of that column
df['LotFrontage'] = df['LotFrontage'].fillna(df['LotFrontage'].mean())
df.head(20)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0    1          60       RL    65.000000     8450   Pave   NaN      Reg   
1    2          20       RL    80.000000     9600   Pave   NaN      Reg   
2    3          60       RL    68.000000    11250   Pave   NaN      IR1   
3    4          70       RL    60.000000     9550   Pave   NaN      IR1   
4    5          60       RL    84.000000    14260   Pave   NaN      IR1   
5    6          50       RL    85.000000    14115   Pave   NaN      IR1   
6    7          20       RL    75.000000    10084   Pave   NaN      Reg   
7    8          60       RL    70.049958    10382   Pave   NaN      IR1   
8    9          50       RM    51.000000     6120   Pave   NaN      Reg   
9   10         190       RL    50.000000     7420   Pave   NaN      Reg   
10  11          20       RL    70.000000    11200   Pave   NaN      Reg   
11  12          60       RL    85.000000    11924   Pave   NaN      IR1   
12  13          20       RL    70.049958    12968   Pave   NaN      IR2   
13  14          20       RL    91.000000    10652   Pave   NaN      IR1   
14  15          20       RL    70.049958    10920   Pave   NaN      IR1   
15  16          45       RM    51.000000     6120   Pave   NaN      Reg   
16  17          20       RL    70.049958    11241   Pave   NaN      IR1   
17  18          90       RL    72.000000    10791   Pave   NaN      Reg   
18  19          20       RL    66.000000    13695   Pave   NaN      Reg   
19  20          20       RL    70.000000     7560   Pave   NaN      Reg   

   LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0          Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1          Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2          Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3          Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4          Lvl    AllPub       FR2       Gtl      NoRidge       Norm   
5          Lvl    AllPub    Inside       Gtl      Mitchel       Norm   
6          Lvl    AllPub    Inside       Gtl      Somerst       Norm   
7          Lvl    AllPub    Corner       Gtl       NWAmes       PosN   
8          Lvl    AllPub    Inside       Gtl      OldTown     Artery   
9          Lvl    AllPub    Corner       Gtl      BrkSide     Artery   
10         Lvl    AllPub    Inside       Gtl       Sawyer       Norm   
11         Lvl    AllPub    Inside       Gtl      NridgHt       Norm   
12         Lvl    AllPub    Inside       Gtl       Sawyer       Norm   
13         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
14         Lvl    AllPub    Corner       Gtl        NAmes       Norm   
15         Lvl    AllPub    Corner       Gtl      BrkSide       Norm   
16         Lvl    AllPub   CulDSac       Gtl        NAmes       Norm   
17         Lvl    AllPub    Inside       Gtl       Sawyer       Norm   
18         Lvl    AllPub    Inside       Gtl      SawyerW       RRAe   
19         Lvl    AllPub    Inside       Gtl        NAmes       Norm   

   Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0        Norm     1Fam     2Story            7            5       2003   
1        Norm     1Fam     1Story            6            8       1976   
2        Norm     1Fam     2Story            7            5       2001   
3        Norm     1Fam     2Story            7            5       1915   
4        Norm     1Fam     2Story            8            5       2000   
5        Norm     1Fam     1.5Fin            5            5       1993   
6        Norm     1Fam     1Story            8            5       2004   
7        Norm     1Fam     2Story            7            6       1973   
8        Norm     1Fam     1.5Fin            7            5       1931   
9      Artery   2fmCon     1.5Unf            5            6       1939   
10       Norm     1Fam     1Story            5            5       1965   
11       Norm     1Fam     

In [8]:
df.drop(columns = 'Alley',inplace = True)
df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street LotShape LandContour  \
0   1          60       RL         65.0     8450   Pave      Reg         Lvl   
1   2          20       RL         80.0     9600   Pave      Reg         Lvl   
2   3          60       RL         68.0    11250   Pave      IR1         Lvl   
3   4          70       RL         60.0     9550   Pave      IR1         Lvl   
4   5          60       RL         84.0    14260   Pave      IR1         Lvl   

  Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType  \
0    AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
1    AllPub       FR2       Gtl      Veenker      Feedr       Norm     1Fam   
2    AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
3    AllPub    Corner       Gtl      Crawfor       Norm       Norm     1Fam   
4    AllPub       FR2       Gtl      NoRidge       Norm       Norm     1Fam   

  HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle  \
0     2Story            7            5       2003          2003     Gable   
1     1Story            6            8       1976          1976     Gable   
2     2Story            7            5       2001          2002     Gable   
3     2Story            7            5       1915          1970     Gable   
4     2Story            8            5       2000          2000     Gable   

  RoofMatl Exterior1st Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond  \
0  CompShg     VinylSd     VinylSd    BrkFace       196.0        Gd        TA   
1  CompShg     MetalSd     MetalSd        NaN         0.0        TA        TA   
2  CompShg     VinylSd     VinylSd    BrkFace       162.0        Gd        TA   
3  CompShg     Wd Sdng     Wd Shng        NaN         0.0        TA        TA   
4  CompShg     VinylSd     VinylSd    BrkFace       350.0        Gd        TA   

  Foundation BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1  \
0      PConc       Gd       TA           No          GLQ         706   
1     CBlock       Gd       TA           Gd          ALQ         978   
2      PConc       Gd       TA           Mn          GLQ         486   
3     BrkTil       TA       Gd           No          ALQ         216   
4      PConc       Gd       TA           Av          GLQ         655   

  BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC  \
0          Unf           0        150          856    GasA        Ex   
1          Unf           0        284         1262    GasA        Ex   
2          Unf           0        434          920    GasA        Ex   
3          Unf           0        540          756    GasA        Gd   
4          Unf           0        490         1145    GasA        Ex   

  CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  \
0          Y      SBrkr       856       854             0       1710   
1          Y      SBrkr      1262         0             0       1262   
2          Y      SBrkr       920       866             0       1786   
3          Y      SBrkr       961       756             0       1717   
4          Y      SBrkr      1145      1053             0       2198   

   BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  \
0             1             0         2         1             3             1   
1             0             1         2         0             3             1   
2             1             0         2         1             3             1   
3             1             0         1         0             3             1   
4             1             0         2         1             4             1   

  KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu GarageType  \
0          Gd             8        Typ           0         NaN     Attchd   
1          TA             6        Typ           1          TA     Attchd   
2          Gd             6        Typ           1          TA     Attchd   
3  

In [9]:
df.drop(columns = 'MasVnrType' ,inplace = True)
df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street LotShape LandContour  \
0   1          60       RL         65.0     8450   Pave      Reg         Lvl   
1   2          20       RL         80.0     9600   Pave      Reg         Lvl   
2   3          60       RL         68.0    11250   Pave      IR1         Lvl   
3   4          70       RL         60.0     9550   Pave      IR1         Lvl   
4   5          60       RL         84.0    14260   Pave      IR1         Lvl   

  Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType  \
0    AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
1    AllPub       FR2       Gtl      Veenker      Feedr       Norm     1Fam   
2    AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
3    AllPub    Corner       Gtl      Crawfor       Norm       Norm     1Fam   
4    AllPub       FR2       Gtl      NoRidge       Norm       Norm     1Fam   

  HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle  \
0     2Story            7            5       2003          2003     Gable   
1     1Story            6            8       1976          1976     Gable   
2     2Story            7            5       2001          2002     Gable   
3     2Story            7            5       1915          1970     Gable   
4     2Story            8            5       2000          2000     Gable   

  RoofMatl Exterior1st Exterior2nd  MasVnrArea ExterQual ExterCond Foundation  \
0  CompShg     VinylSd     VinylSd       196.0        Gd        TA      PConc   
1  CompShg     MetalSd     MetalSd         0.0        TA        TA     CBlock   
2  CompShg     VinylSd     VinylSd       162.0        Gd        TA      PConc   
3  CompShg     Wd Sdng     Wd Shng         0.0        TA        TA     BrkTil   
4  CompShg     VinylSd     VinylSd       350.0        Gd        TA      PConc   

  BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
0       Gd       TA           No          GLQ         706          Unf   
1       Gd       TA           Gd          ALQ         978          Unf   
2       Gd       TA           Mn          GLQ         486          Unf   
3       TA       Gd           No          ALQ         216          Unf   
4       Gd       TA           Av          GLQ         655          Unf   

   BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir Electrical  \
0           0        150          856    GasA        Ex          Y      SBrkr   
1           0        284         1262    GasA        Ex          Y      SBrkr   
2           0        434          920    GasA        Ex          Y      SBrkr   
3           0        540          756    GasA        Gd          Y      SBrkr   
4           0        490         1145    GasA        Ex          Y      SBrkr   

   1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  \
0       856       854             0       1710             1             0   
1      1262         0             0       1262             0             1   
2       920       866             0       1786             1             0   
3       961       756             0       1717             1             0   
4      1145      1053             0       2198             1             0   

   FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
0         2         1             3             1          Gd             8   
1         2         0             3             1          TA             6   
2         2         1             3             1          Gd             6   
3         1         0             3             1          Gd             7   
4         2         1             4             1          Gd             9   

  Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
0        Typ           0         NaN     Attchd       2003.0          RFn   
1        Typ           1          TA     Attchd       1976.0       

In [10]:
df.drop(columns = 'PoolQC',inplace = True)
df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street LotShape LandContour  \
0   1          60       RL         65.0     8450   Pave      Reg         Lvl   
1   2          20       RL         80.0     9600   Pave      Reg         Lvl   
2   3          60       RL         68.0    11250   Pave      IR1         Lvl   
3   4          70       RL         60.0     9550   Pave      IR1         Lvl   
4   5          60       RL         84.0    14260   Pave      IR1         Lvl   

  Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType  \
0    AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
1    AllPub       FR2       Gtl      Veenker      Feedr       Norm     1Fam   
2    AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
3    AllPub    Corner       Gtl      Crawfor       Norm       Norm     1Fam   
4    AllPub       FR2       Gtl      NoRidge       Norm       Norm     1Fam   

  HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle  \
0     2Story            7            5       2003          2003     Gable   
1     1Story            6            8       1976          1976     Gable   
2     2Story            7            5       2001          2002     Gable   
3     2Story            7            5       1915          1970     Gable   
4     2Story            8            5       2000          2000     Gable   

  RoofMatl Exterior1st Exterior2nd  MasVnrArea ExterQual ExterCond Foundation  \
0  CompShg     VinylSd     VinylSd       196.0        Gd        TA      PConc   
1  CompShg     MetalSd     MetalSd         0.0        TA        TA     CBlock   
2  CompShg     VinylSd     VinylSd       162.0        Gd        TA      PConc   
3  CompShg     Wd Sdng     Wd Shng         0.0        TA        TA     BrkTil   
4  CompShg     VinylSd     VinylSd       350.0        Gd        TA      PConc   

  BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
0       Gd       TA           No          GLQ         706          Unf   
1       Gd       TA           Gd          ALQ         978          Unf   
2       Gd       TA           Mn          GLQ         486          Unf   
3       TA       Gd           No          ALQ         216          Unf   
4       Gd       TA           Av          GLQ         655          Unf   

   BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir Electrical  \
0           0        150          856    GasA        Ex          Y      SBrkr   
1           0        284         1262    GasA        Ex          Y      SBrkr   
2           0        434          920    GasA        Ex          Y      SBrkr   
3           0        540          756    GasA        Gd          Y      SBrkr   
4           0        490         1145    GasA        Ex          Y      SBrkr   

   1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  \
0       856       854             0       1710             1             0   
1      1262         0             0       1262             0             1   
2       920       866             0       1786             1             0   
3       961       756             0       1717             1             0   
4      1145      1053             0       2198             1             0   

   FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
0         2         1             3             1          Gd             8   
1         2         0             3             1          TA             6   
2         2         1             3             1          Gd             6   
3         1         0             3             1          Gd             7   
4         2         1             4             1          Gd             9   

  Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
0        Typ           0         NaN     Attchd       2003.0          RFn   
1        Typ           1          TA     Attchd       1976.0       

In [11]:
df.drop(columns = 'Fence',inplace = True)


In [12]:
df.drop(columns = 'MiscFeature',inplace = True)


In [13]:
df['FireplaceQu'] = df['FireplaceQu'].fillna('None')

In [14]:
df['BsmtQual'] = df['BsmtQual'].fillna('None')

In [15]:
df['BsmtCond'] = df['BsmtCond'].fillna('None')
df.head(20)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street LotShape LandContour  \
0    1          60       RL    65.000000     8450   Pave      Reg         Lvl   
1    2          20       RL    80.000000     9600   Pave      Reg         Lvl   
2    3          60       RL    68.000000    11250   Pave      IR1         Lvl   
3    4          70       RL    60.000000     9550   Pave      IR1         Lvl   
4    5          60       RL    84.000000    14260   Pave      IR1         Lvl   
5    6          50       RL    85.000000    14115   Pave      IR1         Lvl   
6    7          20       RL    75.000000    10084   Pave      Reg         Lvl   
7    8          60       RL    70.049958    10382   Pave      IR1         Lvl   
8    9          50       RM    51.000000     6120   Pave      Reg         Lvl   
9   10         190       RL    50.000000     7420   Pave      Reg         Lvl   
10  11          20       RL    70.000000    11200   Pave      Reg         Lvl   
11  12          60       RL    85.000000    11924   Pave      IR1         Lvl   
12  13          20       RL    70.049958    12968   Pave      IR2         Lvl   
13  14          20       RL    91.000000    10652   Pave      IR1         Lvl   
14  15          20       RL    70.049958    10920   Pave      IR1         Lvl   
15  16          45       RM    51.000000     6120   Pave      Reg         Lvl   
16  17          20       RL    70.049958    11241   Pave      IR1         Lvl   
17  18          90       RL    72.000000    10791   Pave      Reg         Lvl   
18  19          20       RL    66.000000    13695   Pave      Reg         Lvl   
19  20          20       RL    70.000000     7560   Pave      Reg         Lvl   

   Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType  \
0     AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
1     AllPub       FR2       Gtl      Veenker      Feedr       Norm     1Fam   
2     AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
3     AllPub    Corner       Gtl      Crawfor       Norm       Norm     1Fam   
4     AllPub       FR2       Gtl      NoRidge       Norm       Norm     1Fam   
5     AllPub    Inside       Gtl      Mitchel       Norm       Norm     1Fam   
6     AllPub    Inside       Gtl      Somerst       Norm       Norm     1Fam   
7     AllPub    Corner       Gtl       NWAmes       PosN       Norm     1Fam   
8     AllPub    Inside       Gtl      OldTown     Artery       Norm     1Fam   
9     AllPub    Corner       Gtl      BrkSide     Artery     Artery   2fmCon   
10    AllPub    Inside       Gtl       Sawyer       Norm       Norm     1Fam   
11    AllPub    Inside       Gtl      NridgHt       Norm       Norm     1Fam   
12    AllPub    Inside       Gtl       Sawyer       Norm       Norm     1Fam   
13    AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
14    AllPub    Corner       Gtl        NAmes       Norm       Norm     1Fam   
15    AllPub    Corner       Gtl      BrkSide       Norm       Norm     1Fam   
16    AllPub   CulDSac       Gtl        NAmes       Norm       Norm     1Fam   
17    AllPub    Inside       Gtl       Sawyer       Norm       Norm   Duplex   
18    AllPub    Inside       Gtl      SawyerW       RRAe       Norm     1Fam   
19    AllPub    Inside       Gtl        NAmes       Norm       Norm     1Fam   

   HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle  \
0      2Story            7            5       2003          2003     Gable   
1      1Story            6            8       1976          1976     Gable   
2      2Story            7            5       2001          2002     Gable   
3      2Story            7            5       1915          1970     Gable   
4      2Story            8            5       2000          2000     Gable   
5      1.5Fin            5            5       1993          1995     Gable   
6      1Story            8            5       2004          2005     Gable 

In [16]:
df['BsmtExposure'] = df['BsmtExposure'].fillna('None')
df.head(20)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street LotShape LandContour  \
0    1          60       RL    65.000000     8450   Pave      Reg         Lvl   
1    2          20       RL    80.000000     9600   Pave      Reg         Lvl   
2    3          60       RL    68.000000    11250   Pave      IR1         Lvl   
3    4          70       RL    60.000000     9550   Pave      IR1         Lvl   
4    5          60       RL    84.000000    14260   Pave      IR1         Lvl   
5    6          50       RL    85.000000    14115   Pave      IR1         Lvl   
6    7          20       RL    75.000000    10084   Pave      Reg         Lvl   
7    8          60       RL    70.049958    10382   Pave      IR1         Lvl   
8    9          50       RM    51.000000     6120   Pave      Reg         Lvl   
9   10         190       RL    50.000000     7420   Pave      Reg         Lvl   
10  11          20       RL    70.000000    11200   Pave      Reg         Lvl   
11  12          60       RL    85.000000    11924   Pave      IR1         Lvl   
12  13          20       RL    70.049958    12968   Pave      IR2         Lvl   
13  14          20       RL    91.000000    10652   Pave      IR1         Lvl   
14  15          20       RL    70.049958    10920   Pave      IR1         Lvl   
15  16          45       RM    51.000000     6120   Pave      Reg         Lvl   
16  17          20       RL    70.049958    11241   Pave      IR1         Lvl   
17  18          90       RL    72.000000    10791   Pave      Reg         Lvl   
18  19          20       RL    66.000000    13695   Pave      Reg         Lvl   
19  20          20       RL    70.000000     7560   Pave      Reg         Lvl   

   Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType  \
0     AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
1     AllPub       FR2       Gtl      Veenker      Feedr       Norm     1Fam   
2     AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
3     AllPub    Corner       Gtl      Crawfor       Norm       Norm     1Fam   
4     AllPub       FR2       Gtl      NoRidge       Norm       Norm     1Fam   
5     AllPub    Inside       Gtl      Mitchel       Norm       Norm     1Fam   
6     AllPub    Inside       Gtl      Somerst       Norm       Norm     1Fam   
7     AllPub    Corner       Gtl       NWAmes       PosN       Norm     1Fam   
8     AllPub    Inside       Gtl      OldTown     Artery       Norm     1Fam   
9     AllPub    Corner       Gtl      BrkSide     Artery     Artery   2fmCon   
10    AllPub    Inside       Gtl       Sawyer       Norm       Norm     1Fam   
11    AllPub    Inside       Gtl      NridgHt       Norm       Norm     1Fam   
12    AllPub    Inside       Gtl       Sawyer       Norm       Norm     1Fam   
13    AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
14    AllPub    Corner       Gtl        NAmes       Norm       Norm     1Fam   
15    AllPub    Corner       Gtl      BrkSide       Norm       Norm     1Fam   
16    AllPub   CulDSac       Gtl        NAmes       Norm       Norm     1Fam   
17    AllPub    Inside       Gtl       Sawyer       Norm       Norm   Duplex   
18    AllPub    Inside       Gtl      SawyerW       RRAe       Norm     1Fam   
19    AllPub    Inside       Gtl        NAmes       Norm       Norm     1Fam   

   HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle  \
0      2Story            7            5       2003          2003     Gable   
1      1Story            6            8       1976          1976     Gable   
2      2Story            7            5       2001          2002     Gable   
3      2Story            7            5       1915          1970     Gable   
4      2Story            8            5       2000          2000     Gable   
5      1.5Fin            5            5       1993          1995     Gable   
6      1Story            8            5       2004          2005     Gable 

In [17]:
df['BsmtFinType1'] = df['BsmtFinType1'].fillna('None')
df.head(20)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street LotShape LandContour  \
0    1          60       RL    65.000000     8450   Pave      Reg         Lvl   
1    2          20       RL    80.000000     9600   Pave      Reg         Lvl   
2    3          60       RL    68.000000    11250   Pave      IR1         Lvl   
3    4          70       RL    60.000000     9550   Pave      IR1         Lvl   
4    5          60       RL    84.000000    14260   Pave      IR1         Lvl   
5    6          50       RL    85.000000    14115   Pave      IR1         Lvl   
6    7          20       RL    75.000000    10084   Pave      Reg         Lvl   
7    8          60       RL    70.049958    10382   Pave      IR1         Lvl   
8    9          50       RM    51.000000     6120   Pave      Reg         Lvl   
9   10         190       RL    50.000000     7420   Pave      Reg         Lvl   
10  11          20       RL    70.000000    11200   Pave      Reg         Lvl   
11  12          60       RL    85.000000    11924   Pave      IR1         Lvl   
12  13          20       RL    70.049958    12968   Pave      IR2         Lvl   
13  14          20       RL    91.000000    10652   Pave      IR1         Lvl   
14  15          20       RL    70.049958    10920   Pave      IR1         Lvl   
15  16          45       RM    51.000000     6120   Pave      Reg         Lvl   
16  17          20       RL    70.049958    11241   Pave      IR1         Lvl   
17  18          90       RL    72.000000    10791   Pave      Reg         Lvl   
18  19          20       RL    66.000000    13695   Pave      Reg         Lvl   
19  20          20       RL    70.000000     7560   Pave      Reg         Lvl   

   Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType  \
0     AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
1     AllPub       FR2       Gtl      Veenker      Feedr       Norm     1Fam   
2     AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
3     AllPub    Corner       Gtl      Crawfor       Norm       Norm     1Fam   
4     AllPub       FR2       Gtl      NoRidge       Norm       Norm     1Fam   
5     AllPub    Inside       Gtl      Mitchel       Norm       Norm     1Fam   
6     AllPub    Inside       Gtl      Somerst       Norm       Norm     1Fam   
7     AllPub    Corner       Gtl       NWAmes       PosN       Norm     1Fam   
8     AllPub    Inside       Gtl      OldTown     Artery       Norm     1Fam   
9     AllPub    Corner       Gtl      BrkSide     Artery     Artery   2fmCon   
10    AllPub    Inside       Gtl       Sawyer       Norm       Norm     1Fam   
11    AllPub    Inside       Gtl      NridgHt       Norm       Norm     1Fam   
12    AllPub    Inside       Gtl       Sawyer       Norm       Norm     1Fam   
13    AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
14    AllPub    Corner       Gtl        NAmes       Norm       Norm     1Fam   
15    AllPub    Corner       Gtl      BrkSide       Norm       Norm     1Fam   
16    AllPub   CulDSac       Gtl        NAmes       Norm       Norm     1Fam   
17    AllPub    Inside       Gtl       Sawyer       Norm       Norm   Duplex   
18    AllPub    Inside       Gtl      SawyerW       RRAe       Norm     1Fam   
19    AllPub    Inside       Gtl        NAmes       Norm       Norm     1Fam   

   HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle  \
0      2Story            7            5       2003          2003     Gable   
1      1Story            6            8       1976          1976     Gable   
2      2Story            7            5       2001          2002     Gable   
3      2Story            7            5       1915          1970     Gable   
4      2Story            8            5       2000          2000     Gable   
5      1.5Fin            5            5       1993          1995     Gable   
6      1Story            8            5       2004          2005     Gable 

In [18]:
df['BsmtFinType2'] = df['BsmtFinType2'].fillna('None')
df.head(20)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street LotShape LandContour  \
0    1          60       RL    65.000000     8450   Pave      Reg         Lvl   
1    2          20       RL    80.000000     9600   Pave      Reg         Lvl   
2    3          60       RL    68.000000    11250   Pave      IR1         Lvl   
3    4          70       RL    60.000000     9550   Pave      IR1         Lvl   
4    5          60       RL    84.000000    14260   Pave      IR1         Lvl   
5    6          50       RL    85.000000    14115   Pave      IR1         Lvl   
6    7          20       RL    75.000000    10084   Pave      Reg         Lvl   
7    8          60       RL    70.049958    10382   Pave      IR1         Lvl   
8    9          50       RM    51.000000     6120   Pave      Reg         Lvl   
9   10         190       RL    50.000000     7420   Pave      Reg         Lvl   
10  11          20       RL    70.000000    11200   Pave      Reg         Lvl   
11  12          60       RL    85.000000    11924   Pave      IR1         Lvl   
12  13          20       RL    70.049958    12968   Pave      IR2         Lvl   
13  14          20       RL    91.000000    10652   Pave      IR1         Lvl   
14  15          20       RL    70.049958    10920   Pave      IR1         Lvl   
15  16          45       RM    51.000000     6120   Pave      Reg         Lvl   
16  17          20       RL    70.049958    11241   Pave      IR1         Lvl   
17  18          90       RL    72.000000    10791   Pave      Reg         Lvl   
18  19          20       RL    66.000000    13695   Pave      Reg         Lvl   
19  20          20       RL    70.000000     7560   Pave      Reg         Lvl   

   Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType  \
0     AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
1     AllPub       FR2       Gtl      Veenker      Feedr       Norm     1Fam   
2     AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
3     AllPub    Corner       Gtl      Crawfor       Norm       Norm     1Fam   
4     AllPub       FR2       Gtl      NoRidge       Norm       Norm     1Fam   
5     AllPub    Inside       Gtl      Mitchel       Norm       Norm     1Fam   
6     AllPub    Inside       Gtl      Somerst       Norm       Norm     1Fam   
7     AllPub    Corner       Gtl       NWAmes       PosN       Norm     1Fam   
8     AllPub    Inside       Gtl      OldTown     Artery       Norm     1Fam   
9     AllPub    Corner       Gtl      BrkSide     Artery     Artery   2fmCon   
10    AllPub    Inside       Gtl       Sawyer       Norm       Norm     1Fam   
11    AllPub    Inside       Gtl      NridgHt       Norm       Norm     1Fam   
12    AllPub    Inside       Gtl       Sawyer       Norm       Norm     1Fam   
13    AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
14    AllPub    Corner       Gtl        NAmes       Norm       Norm     1Fam   
15    AllPub    Corner       Gtl      BrkSide       Norm       Norm     1Fam   
16    AllPub   CulDSac       Gtl        NAmes       Norm       Norm     1Fam   
17    AllPub    Inside       Gtl       Sawyer       Norm       Norm   Duplex   
18    AllPub    Inside       Gtl      SawyerW       RRAe       Norm     1Fam   
19    AllPub    Inside       Gtl        NAmes       Norm       Norm     1Fam   

   HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle  \
0      2Story            7            5       2003          2003     Gable   
1      1Story            6            8       1976          1976     Gable   
2      2Story            7            5       2001          2002     Gable   
3      2Story            7            5       1915          1970     Gable   
4      2Story            8            5       2000          2000     Gable   
5      1.5Fin            5            5       1993          1995     Gable   
6      1Story            8            5       2004          2005     Gable 

In [19]:
df.isnull().sum()

Id                0
MSSubClass        0
MSZoning          0
LotFrontage       0
LotArea           0
Street            0
LotShape          0
LandContour       0
Utilities         0
LotConfig         0
LandSlope         0
Neighborhood      0
Condition1        0
Condition2        0
BldgType          0
HouseStyle        0
OverallQual       0
OverallCond       0
YearBuilt         0
YearRemodAdd      0
RoofStyle         0
RoofMatl          0
Exterior1st       0
Exterior2nd       0
MasVnrArea        8
ExterQual         0
ExterCond         0
Foundation        0
BsmtQual          0
BsmtCond          0
BsmtExposure      0
BsmtFinType1      0
BsmtFinSF1        0
BsmtFinType2      0
BsmtFinSF2        0
BsmtUnfSF         0
TotalBsmtSF       0
Heating           0
HeatingQC         0
CentralAir        0
Electrical        1
1stFlrSF          0
2ndFlrSF          0
LowQualFinSF      0
GrLivArea         0
BsmtFullBath      0
BsmtHalfBath      0
FullBath          0
HalfBath          0
BedroomAbvGr      0


In [20]:
df['MasVnrArea'] = df['MasVnrArea'].fillna(df['MasVnrArea'].mean())


In [21]:
df['GarageType'] = df['GarageType'].fillna('None')


In [22]:
df['GarageYrBlt'] = df['GarageYrBlt'].fillna(0)


In [23]:
df['GarageFinish'] = df['GarageFinish'].fillna('None')

In [24]:
df['GarageQual'] = df['GarageQual'].fillna('None')

In [25]:
df['GarageCond'] = df['GarageCond'].fillna('None')


In [26]:
df['Electrical'] = df['Electrical'].fillna(df['Electrical'].mode()[0])


In [29]:
df.isnull().sum()

Id               0
MSSubClass       0
MSZoning         0
LotFrontage      0
LotArea          0
Street           0
LotShape         0
LandContour      0
Utilities        0
LotConfig        0
LandSlope        0
Neighborhood     0
Condition1       0
Condition2       0
BldgType         0
HouseStyle       0
OverallQual      0
OverallCond      0
YearBuilt        0
YearRemodAdd     0
RoofStyle        0
RoofMatl         0
Exterior1st      0
Exterior2nd      0
MasVnrArea       0
ExterQual        0
ExterCond        0
Foundation       0
BsmtQual         0
BsmtCond         0
BsmtExposure     0
BsmtFinType1     0
BsmtFinSF1       0
BsmtFinType2     0
BsmtFinSF2       0
BsmtUnfSF        0
TotalBsmtSF      0
Heating          0
HeatingQC        0
CentralAir       0
Electrical       0
1stFlrSF         0
2ndFlrSF         0
LowQualFinSF     0
GrLivArea        0
BsmtFullBath     0
BsmtHalfBath     0
FullBath         0
HalfBath         0
BedroomAbvGr     0
KitchenAbvGr     0
KitchenQual      0
TotRmsAbvGrd

In [30]:
#based on observation  we have to perform feature engineering on the dataset to get better accuracy in machine learning model 


In [31]:
df.head(20)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street LotShape LandContour  \
0    1          60       RL    65.000000     8450   Pave      Reg         Lvl   
1    2          20       RL    80.000000     9600   Pave      Reg         Lvl   
2    3          60       RL    68.000000    11250   Pave      IR1         Lvl   
3    4          70       RL    60.000000     9550   Pave      IR1         Lvl   
4    5          60       RL    84.000000    14260   Pave      IR1         Lvl   
5    6          50       RL    85.000000    14115   Pave      IR1         Lvl   
6    7          20       RL    75.000000    10084   Pave      Reg         Lvl   
7    8          60       RL    70.049958    10382   Pave      IR1         Lvl   
8    9          50       RM    51.000000     6120   Pave      Reg         Lvl   
9   10         190       RL    50.000000     7420   Pave      Reg         Lvl   
10  11          20       RL    70.000000    11200   Pave      Reg         Lvl   
11  12          60       RL    85.000000    11924   Pave      IR1         Lvl   
12  13          20       RL    70.049958    12968   Pave      IR2         Lvl   
13  14          20       RL    91.000000    10652   Pave      IR1         Lvl   
14  15          20       RL    70.049958    10920   Pave      IR1         Lvl   
15  16          45       RM    51.000000     6120   Pave      Reg         Lvl   
16  17          20       RL    70.049958    11241   Pave      IR1         Lvl   
17  18          90       RL    72.000000    10791   Pave      Reg         Lvl   
18  19          20       RL    66.000000    13695   Pave      Reg         Lvl   
19  20          20       RL    70.000000     7560   Pave      Reg         Lvl   

   Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType  \
0     AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
1     AllPub       FR2       Gtl      Veenker      Feedr       Norm     1Fam   
2     AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
3     AllPub    Corner       Gtl      Crawfor       Norm       Norm     1Fam   
4     AllPub       FR2       Gtl      NoRidge       Norm       Norm     1Fam   
5     AllPub    Inside       Gtl      Mitchel       Norm       Norm     1Fam   
6     AllPub    Inside       Gtl      Somerst       Norm       Norm     1Fam   
7     AllPub    Corner       Gtl       NWAmes       PosN       Norm     1Fam   
8     AllPub    Inside       Gtl      OldTown     Artery       Norm     1Fam   
9     AllPub    Corner       Gtl      BrkSide     Artery     Artery   2fmCon   
10    AllPub    Inside       Gtl       Sawyer       Norm       Norm     1Fam   
11    AllPub    Inside       Gtl      NridgHt       Norm       Norm     1Fam   
12    AllPub    Inside       Gtl       Sawyer       Norm       Norm     1Fam   
13    AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
14    AllPub    Corner       Gtl        NAmes       Norm       Norm     1Fam   
15    AllPub    Corner       Gtl      BrkSide       Norm       Norm     1Fam   
16    AllPub   CulDSac       Gtl        NAmes       Norm       Norm     1Fam   
17    AllPub    Inside       Gtl       Sawyer       Norm       Norm   Duplex   
18    AllPub    Inside       Gtl      SawyerW       RRAe       Norm     1Fam   
19    AllPub    Inside       Gtl        NAmes       Norm       Norm     1Fam   

   HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle  \
0      2Story            7            5       2003          2003     Gable   
1      1Story            6            8       1976          1976     Gable   
2      2Story            7            5       2001          2002     Gable   
3      2Story            7            5       1915          1970     Gable   
4      2Story            8            5       2000          2000     Gable   
5      1.5Fin            5            5       1993          1995     Gable   
6      1Story            8            5       2004          2005     Gable 

In [38]:
df['Street'].value_counts(normalize = True)

Street
Pave    0.99589
Grvl    0.00411
Name: proportion, dtype: float64

In [39]:
#since it will cause overfitting removing is best

In [43]:
df.drop(columns = 'Street',axis = 1, inplace = True)

In [42]:
df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea LotShape LandContour  \
0   1          60       RL         65.0     8450      Reg         Lvl   
1   2          20       RL         80.0     9600      Reg         Lvl   
2   3          60       RL         68.0    11250      IR1         Lvl   
3   4          70       RL         60.0     9550      IR1         Lvl   
4   5          60       RL         84.0    14260      IR1         Lvl   

  Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType  \
0    AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
1    AllPub       FR2       Gtl      Veenker      Feedr       Norm     1Fam   
2    AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
3    AllPub    Corner       Gtl      Crawfor       Norm       Norm     1Fam   
4    AllPub       FR2       Gtl      NoRidge       Norm       Norm     1Fam   

  HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle  \
0     2Story            7            5       2003          2003     Gable   
1     1Story            6            8       1976          1976     Gable   
2     2Story            7            5       2001          2002     Gable   
3     2Story            7            5       1915          1970     Gable   
4     2Story            8            5       2000          2000     Gable   

  RoofMatl Exterior1st Exterior2nd  MasVnrArea ExterQual ExterCond Foundation  \
0  CompShg     VinylSd     VinylSd       196.0        Gd        TA      PConc   
1  CompShg     MetalSd     MetalSd         0.0        TA        TA     CBlock   
2  CompShg     VinylSd     VinylSd       162.0        Gd        TA      PConc   
3  CompShg     Wd Sdng     Wd Shng         0.0        TA        TA     BrkTil   
4  CompShg     VinylSd     VinylSd       350.0        Gd        TA      PConc   

  BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
0       Gd       TA           No          GLQ         706          Unf   
1       Gd       TA           Gd          ALQ         978          Unf   
2       Gd       TA           Mn          GLQ         486          Unf   
3       TA       Gd           No          ALQ         216          Unf   
4       Gd       TA           Av          GLQ         655          Unf   

   BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir Electrical  \
0           0        150          856    GasA        Ex          Y      SBrkr   
1           0        284         1262    GasA        Ex          Y      SBrkr   
2           0        434          920    GasA        Ex          Y      SBrkr   
3           0        540          756    GasA        Gd          Y      SBrkr   
4           0        490         1145    GasA        Ex          Y      SBrkr   

   1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  \
0       856       854             0       1710             1             0   
1      1262         0             0       1262             0             1   
2       920       866             0       1786             1             0   
3       961       756             0       1717             1             0   
4      1145      1053             0       2198             1             0   

   FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd  \
0         2         1             3             1          Gd             8   
1         2         0             3             1          TA             6   
2         2         1             3             1          Gd             6   
3         1         0             3             1          Gd             7   
4         2         1             4             1          Gd             9   

  Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  \
0        Typ           0        None     Attchd       2003.0          RFn   
1        Typ           1          TA     Attchd       1976.0          RFn   
2        Typ           1        

In [46]:
df.dtypes

Id                 int64
MSSubClass         int64
MSZoning          object
LotFrontage      float64
LotArea            int64
LotShape          object
LandContour       object
Utilities         object
LotConfig         object
LandSlope         object
Neighborhood      object
Condition1        object
Condition2        object
BldgType          object
HouseStyle        object
OverallQual        int64
OverallCond        int64
YearBuilt          int64
YearRemodAdd       int64
RoofStyle         object
RoofMatl          object
Exterior1st       object
Exterior2nd       object
MasVnrArea       float64
ExterQual         object
ExterCond         object
Foundation        object
BsmtQual          object
BsmtCond          object
BsmtExposure      object
BsmtFinType1      object
BsmtFinSF1         int64
BsmtFinType2      object
BsmtFinSF2         int64
BsmtUnfSF          int64
TotalBsmtSF        int64
Heating           object
HeatingQC         object
CentralAir        object
Electrical        object
